### PR TITLE
Critical Hit / Overkill damage overhaul

### PR DESCRIPTION
**Pull request by TranslucentSabre**
# Description

These changes overhaul how critical hits and overkill heat is calculated and displayed. The new method is more accurate for critical hit damage, and ( subjectively ) displays roll output more clearly. I will provide a summary of the algorithm changes, with screenshots, in the comments.

It also now accounts for overkill with bonus damage as well.

## Issue Number
`#1404`

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by TranslucentSabre** [comment]
Algorithm changes:
For our examples, let's use a hypothetical damage roll of 3d4.

The old algorithm would roll two different sets of 3d4 ( re-rolling overkill 1s ), get the sum for each set ( including overkill 1s! ) and use whichever set had the highest total. The values from the "lower" set would be displayed but not used on the Dice panel. Please see this example:
image

This is already not how the book describes critical hits, but has several other flaws as well. 
1) Overkill re-rolls in the "lower" set are not counted as overkill re-rolls and do not generate heat.
2) Since overkill 1s _are_ counted when figuring out which set is "higher" a large number of overkill can and will cause the set with the lower real damage to be selected. ( This is seen in the screenshot above )
3) Due to the above, the UI can make it seem like overkill 1s are re-rolled instead of taking the higher values.

-------------------------------------------------------------------------------------------------------------------------------

The new algorithm attempts to follow the description in the book as exactly as possible. It rolls 2 times the number of dice, re-rolling overkill 1s until we have 2 times the number of dice non-1 values. Then we iterate over the values, marking overkill 1s and the high values to be used for damage. This is displayed as one list of rolls in the UI.
Please see this example:
image


These changes pass all existing unit tests run by `yarn unit-test`.

File: src/classes/dice/DiceRoller.ts [hunk 1]
```diff
@@ -206,41 +206,57 @@ class DiceRoller {
...
+    
+    const rollClass: string[] = []
+    //Set initial 'overkill' and 'low' values
+    rolls.forEach((value, index) => {
+      if (overkill && (value === 1)) {
+        rollClass.push('overkill')
+      } else {
+        rollClass.push('low')
+      }
+    })
```
Comment (TranslucentSabre) [inline src/classes/dice/DiceRoller.ts:240]: I realized after pushing that this whole forEach to set the initial value of rollClass can be a much smaller roll.map() call. Would you like me to make that change?

File: src/classes/dice/DiceRoller.ts [hunk 2]
```diff
@@ -206,41 +206,57 @@ class DiceRoller {
...
+      rolls.forEach((value, index) => {
+        if(value > highValue && rollClass[index] != 'high') {
+          highValue = value
+          highIndex = index
+        }
+      })
+      if(highValue > 0) {
+        rollClass[highIndex] = 'high'
+      }
+    }
```
Comment (TranslucentSabre) [inline src/classes/dice/DiceRoller.ts:255]: This loop too, I'm not very used to thinking or using functional programming concepts, so if there is a better / more consistent way to find and mark the highest unique `n` values, please let me know and I'll happily change it.

File: src/classes/dice/DiceRoller.ts [hunk 3]
```diff
@@ -270,14 +286,15 @@ class DiceRoller {
     }
   }
 
-  public static rollDieSet(dieSet: DieSet, overkill?: boolean): { result: number; rolls: number[]; rerolls: number } {
+  public static rollDieSet(dieSet: DieSet, overkill?: boolean, critical?: boolean): { result: number; rolls: number[]; rerolls: number } {
     if (dieSet.quantity <= 0 || dieSet.type <= 0) return { result: 0, rolls: [], rerolls: 0 }
 
     let total = 0
     let rerolls = 0
+    let quantity = critical ? dieSet.quantity * 2 : dieSet.quantity
```
Comment (TranslucentSabre) [inline src/classes/dice/DiceRoller.ts:294]: This should probably be const, shouldn't it?

File: src/classes/dice/DiceRoller.ts (hunk 1 above)
Comment (jarena3) [reply src/classes/dice/DiceRoller.ts:240]: I think this code (and the other comments here) are perfectly fine and readable, this (and the other ideas you've marked) are totally at your discretion; I'm good either way

**Review by jarena3 — APPROVED** [review]

**Pull request merged**